flow/layers/container_layer.cc

```diff
@@ -10,6 +10,83 @@ namespace flutter {
 
 ContainerLayer::ContainerLayer() {}
 
+void ContainerLayer::AssignOldLayer(std::shared_ptr<ContainerLayer> old_layer) {
+  old_layer->old_layer_ = nullptr;
+  old_layer_ = old_layer;
+}
+
+void ContainerLayer::Diff(DiffContext* context, const Layer* old_layer) {
+  auto old_container = reinterpret_cast<const ContainerLayer*>(old_layer);
+  auto tree = context->BeginSubtree();
+  DiffChildren(context, old_container);
+  set_paint_region(context->CurrentSubtreeRegion());
+}
+
+void ContainerLayer::DiffChildren(DiffContext* context,
+                                  const ContainerLayer* old_layer) {
+  if (context->IsSubtreeDirty()) {
+    for (auto& layer : layers_) {
+      layer->Diff(context, nullptr);
+    }
+  } else {
+    assert(old_layer);
+
+    const auto& prev_layers = old_layer->layers_;
+
+    int first_changed, end, last_cur, last_prev;
+    end = layers_.size();
+    first_changed = 0;
+    last_cur = end;
+    last_prev = prev_layers.size();
+    while (first_changed < last_cur && first_changed < last_prev &&
+           layers_[first_changed]->CanDiff(prev_layers[first_changed].get())) {
+      first_changed++;
+    }
+    while (first_changed < last_cur - 1 && first_changed < last_prev - 1 &&
+           layers_[last_cur - 1]->CanDiff(prev_layers[last_prev - 1].get())) {
+      last_cur--;
+      last_prev--;
+    }
+
+    // old layers that don't match
+    for (int i = first_changed; i < last_prev; ++i) {
+      auto layer = prev_layers[i];
+      auto subtree = context->BeginSubtree();
+      context->MarkSubtreeDirty(layer->paint_region());
+    }
+
+    for (int i = 0; i < end; ++i) {
+      if (i < first_changed) {
```

Comment:
That was actually taken straight out of @flar's prototype :) Noted.

Reply:
I have to wrap my head around this again. ;)

Another option is to use 3 separate loops to avoid having to do a lot of tests on `i` in each iteration.
```
for (int i = 0; i < first_changed; i++) ...
for (int i = first_changed; i < last_cur; i++) ...
for (int i = last_cur, i_prev = ...; i < end; i++, i_prev--) ...
```
The first and last loops are very similar, but we avoid the calculation of `i_prev` on every iteration and avoid 2-3 tests on `i` in many cases.

Reply:
I rewrote the code to be more in line with `Element.updateChildren`, using `new_children_top`, `new_children_bottom`, `old_children_top` and `old_children_bottom` to represent the mismatched elements. I'll create new PR soon.

Reply:
(And I just realized that I should have been writing `i_prev++` in those past couple of comments. D'oh!)